File: max-sub-array-sum/kadanes_algo.js
/**
 *
 * @param {number} nums
 * @returns {number}
 * @description Takes in an array of integers and returns the sum of the largest continuos subarray
 */
const maxSubarrSum = (nums) => {
  let maxSum = Number.MIN_SAFE_INTEGER;
  let curSum = 0;

  for (let i = 0; i < nums.length; i++) {
    // Capture the current element
    const cur = nums[i];

    // set curSum to the larger between current element or previous curSum + current element
    curSum = Math.max(cur, curSum + cur);

    // set maxSum (result) to the larger between current maxSum or curSum
    maxSum = Math.max(maxSum, curSum);
  }

  return maxSum;
};

console.log("Max Subarr Sum ---> ", maxSubarrSum([-1, 1, 2, 5, -8, 6])); // 8
console.log("Max Subarr Sum ---> ", maxSubarrSum([-1, 1, 2, 5, -8, 9])); // 9
console.log("Max Subarr Sum ---> ", maxSubarrSum([-1, 1, 2, 5 - 5, 6])); // 9
console.log("Max Subarr Sum ---> ", maxSubarrSum([-1, 1, 2, 5, -3, 6])); // 11

File: max-sub-array-sum/kadanes_algo.py
from typing import List
def maxSubarraySum(nums: List[int]) -> int:
    maxSum = float('-inf')
    curSum = 0

    for n in nums:
        curSum = max(n, curSum+n)
        maxSum = max(curSum, maxSum)
    
    return maxSum

print('Kadanes algo --> ', maxSubarraySum([-1,2,4,-1,-1,5,5])) # 14
print('Kadanes algo --> ', maxSubarraySum([2,5,-1,-1,-1,-1,-1,-1,-1,-1,2,3])) # 7
print('Kadanes algo --> ', maxSubarraySum([2,2,-1,-1,-1,-1,-1,-1,-1,-1,2,3])) # 5
